### ISSUE TITLE
Regression with (un)maximizing Neovim in a terminal window

### ISSUE THREAD

**Issue by blueyed** [neovim/neovim#3385]
Commit 47cbbc0b8c864bdb1aaa572f0396642e2c66deae causes a regression (again, see #2322) when maximizing a floating terminal window with Neovim in it: the width does not seem to be updated correctly.  That might depend on how the window manager does the maximizing - a previous workaround has been to "slowly" do first the vertical and then the horizontal maximization.

Some more info from bisecting: with e897cca it works when maximizing, but has artifacts when restoring it again. Same for 47cbbc0.

The commit references issue 3332, but that is likely a typo.

**Comment by tarruda** [comment]
I find it strange that @47cbbc0 has anything to do with this since it doesn't touch the terminal UI(which is what the #2322 issue refers to). Also, Neovim doesn't have any knowledge if the terminal window is floating or not, it just handles SIGWINCH signals by querying the new size with `uv_tty_get_winsize`. 

What is printed if you apply this patch and redirect stderr to a file?

```diff
diff --git a/src/nvim/tui/tui.c b/src/nvim/tui/tui.c
index 4c8e88c..2005e5c 100644
--- a/src/nvim/tui/tui.c
+++ b/src/nvim/tui/tui.c
@@ -678,8 +678,10 @@ static void update_size(UI *ui)

   // 2 - try from a system call(ioctl/TIOCGWINSZ on unix)
   if (!uv_tty_get_winsize(&data->output_handle, &width, &height)) {
+    fprintf(stderr, "failed to get new size\n");
     goto end;
   }
+  fprintf(stderr, "new size is %d/%d\n", width, height);

   // 3 - use $LINES/$COLUMNS if available
   const char *val;
```

**Comment by blueyed** [comment]
(47cbbc0 is not really related, it was just a (also bad) step during bisecting)

The patch prints `failed to get new size` only.
For (un)maximizing it gets printed three times.  Otherwise only once.

**Comment by justinmk** [comment]
If `uv_tty_get_winsize` returns 0, it means "success", according to http://docs.libuv.org/en/v1.x/tty.html#c.uv_tty_get_winsize . Not `"failed to get new size\n"`

**Comment by blueyed** [comment]
I see, updated the patch to switch the print statements:

new size is 84/32
Maximizize:
new size is 84/43
new size is 168/43
new size is 169/43
Restore:
new size is 169/32
new size is 168/32
new size is 84/32

It seems to look better / different from yesterday already: "only" the cmdwin height, and the display of the tabline seems to be affected mainly.  But the width of the window seems to be OK in general (at least it gets adopted).
When doing this multiple times, there might be artifacts in the lower right of the window.

**Comment by tarruda** [comment]
I applied the `uv_tty_get_winsize` fix to #3360

**Comment by tarruda** [comment]
> If uv_tty_get_winsize returns 0, it means "success", according to http://docs.libuv.org/en/v1.x/tty.html#c.uv_tty_get_winsize . Not "failed to get new size\n".

Strangely enough, on linux it seems to be returning 0 for failure, which I believe is the reason I added this condition in the first place. Need to investigate

**Comment by tarruda** [comment]
@justinmk right, I had it all wrong. I understood your comment as a hint to fix the tui.c while in truth it was to fix the patch I sent @blueyed . Sorry for the complete confusion :confounded:

**Comment by tarruda** [comment]
@blueyed so after you switched the print statements the output is as expected, but the size still isn't updated in the terminal window?

**Comment by justinmk** [comment]
> a previous workaround has been to "slowly" do first the vertical and then the horizontal maximization.

@blueyed Is that workaround in Vim? Or what do you mean?

**Comment by blueyed** [comment]
> @blueyed so after you switched the print statements the output is as expected, but the size still isn't updated in the terminal window?

It's not updated properly, yes.

> @blueyed Is that workaround in Vim? Or what do you mean?

It's a workaround with the window manager.
I can imagine that Neovim and awesome window manager might be required to trigger this behaviour (because of the SIGWINCH handling).

**Comment by blueyed** [comment]
I've noticed now that my workaround of first maximizing it vertically and then horizontally fails with the following:

    au VimResized * sleep 100m

With this `sleep` there will be only one VimResized event being triggered (instead of ~3) otherwise.

See:

    au VimResized *         echom i | let i = i+1 | sleep 100m

`printf` from the patch (also there with the `sleep`):

Maximize:
new size is 94/52
new size is 94/51
new size is 190/51
new size is 190/52

Un-maximize:
new size is 94/52

Maximize:
new size is 94/51
new size is 190/51
new size is 190/52

```diff
 src/nvim/tui/tui.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git i/src/nvim/tui/tui.c w/src/nvim/tui/tui.c
index 00e2821..8165ad2 100644
--- i/src/nvim/tui/tui.c
+++ w/src/nvim/tui/tui.c
@@ -696,8 +698,10 @@ static void update_size(UI *ui)
   // 2 - try from a system call(ioctl/TIOCGWINSZ on unix)
   if (data->out_isatty &&
       !uv_tty_get_winsize(&data->output_handle.tty, &width, &height)) {
+    fprintf(stderr, "new size is %d/%d\n", width, height);
     goto end;
   }
+  fprintf(stderr, "failed to get new size\n");
 
   // 3 - use $LINES/$COLUMNS if available
   const char *val;
```

Neovim bf2c2b3 (master).

**Comment by tartley** [comment]
I see a similar sounding problem with Neovim NVIM 0.2.0-dev in gnome-terminal on a vanilla Ubuntu 16.04 desktop, if that's a useful data point.

About 7 times out of 8, hitting F11 to fullscreen a floating Neovim window doesn't update the height and width of neovim - it is the same size, but now occupies the top left of the fullscreened window, with black bars on the right and bottom filling the rest of the screen. The other 1 out of 8 times, fullscreening works fine.

Fullscreening a plain gnome-terminal works fine.

Apologies for the noise if this is all obvious or irrelevant.

**Comment by justinmk** [comment]
fixed by 34e5654f2166 and https://github.com/neovim/neovim/pull/6202

@blueyed can you confirm?

**Comment by vvbalashoff** [comment]
On Gentoo x86_64, nvim 0.3.4 in fullscreen runs fine.
But I always run my roxterm in window mode, when I can see task bar.
So, when I run nvim, it collapse roxterm window with all it's tabs to one shot line, where I can see:
===
[No Name]   0,0-1   All
===
and that's all. I have to resize my roxterm window manually every time after running nvim.
